### PR TITLE
Face contact

### PR DESCRIPTION
**Pull request by xhan0619**
Add support for:

1. Separate simulation and render mesh.
2. Deformable face vs. rigid contact.
3. Deformable face vs. deformable node contact.
4. Deformable self-collision

Added more documentation.

### CODE REVIEW

**Comment by erwincoumans** [comment]
Thanks a lot for the pull request, I will look at it.

File: src/BulletCollision/BroadphaseCollision/btDbvt.h [hunk 1]
```diff
@@ -21,7 +21,7 @@ subject to the following restrictions:
 #include "LinearMath/btVector3.h"
 #include "LinearMath/btTransform.h"
 #include "LinearMath/btAabbUtil2.h"
-
+#include <iostream>
```
Comment (erwincoumans) [inline src/BulletCollision/BroadphaseCollision/btDbvt.h]: It is not needed, right? Please remove iostream.

File: src/BulletSoftBody/btDeformableMultiBodyDynamicsWorld.cpp [hunk 2]
```diff
@@ -66,86 +75,21 @@ void btDeformableMultiBodyDynamicsWorld::internalSingleStepSimulation(btScalar t
...
 }
 
-
 void btDeformableMultiBodyDynamicsWorld::integrateTransforms(btScalar timeStep)
 {
     BT_PROFILE("integrateTransforms");
-    m_deformableBodySolver->backupVelocity();
-    positionCorrection(timeStep);
+    //m_deformableBodySolver->backupVelocity();
+    //positionCorrection(timeStep);  // looks like position correction is no longer necessary
```
Comment (erwincoumans) [inline src/BulletSoftBody/btDeformableMultiBodyDynamicsWorld.cpp]: Why? Is position correction applied somewhere else?

File: src/BulletCollision/BroadphaseCollision/btDbvt.h (hunk 1 above)
Comment (xhan0619) [reply src/BulletCollision/BroadphaseCollision/btDbvt.h]: Done.

File: src/BulletSoftBody/btDeformableMultiBodyDynamicsWorld.cpp (hunk 2 above)
Comment (xhan0619) [reply src/BulletSoftBody/btDeformableMultiBodyDynamicsWorld.cpp]: Position correction was introduced to combat drift in deformable related contact. From the results of my experiments, drift is no longer visible. The position correction is also introducing instability when the deformable object is stiff.

File: src/BulletSoftBody/btDeformableBodySolver.cpp [hunk 3]
```diff
@@ -79,31 +104,101 @@ void btDeformableBodySolver::solveDeformableConstraints(btScalar solverdt)
...
+
+btScalar btDeformableBodySolver::computeDescentStep(TVStack& ddv, const TVStack& residual)
+{
+    m_cg.solve(*m_objective, ddv, residual, false);
+    btScalar inner_product = m_cg.dot(residual, m_ddv);
+    btScalar res_norm = m_objective->computeNorm(residual);
+    btScalar tol = 1e-5 * res_norm * m_objective->computeNorm(m_ddv);
+    if (inner_product < -tol)
+    {
+        std::cout << "Looking backwards!" << std::endl;
```
Comment (erwincoumans) [inline src/BulletSoftBody/btDeformableBodySolver.cpp]: add a 'verbose' so we don't print this?

File: src/BulletSoftBody/btDeformableBodySolver.cpp [hunk 4]
```diff
@@ -79,31 +104,101 @@ void btDeformableBodySolver::solveDeformableConstraints(btScalar solverdt)
...
+        for (int i = 0; i < m_ddv.size();++i)
+        {
+            m_ddv[i] = -m_ddv[i];
+        }
+        inner_product = -inner_product;
+    }
+    else if (std::abs(inner_product) < tol)
+    {
+        std::cout << "Gradient Descent!" << std::endl;
```
Comment (erwincoumans) [inline src/BulletSoftBody/btDeformableBodySolver.cpp]: conditional 'verbose'?

File: src/BulletCollision/BroadphaseCollision/btDbvt.h [hunk 5]
```diff
@@ -190,6 +192,36 @@ struct btDbvtNode
...
+    , normal(0,0,0)
+    , data(n->data)
+    {
+        childs[0] = 0;
+        childs[1] = 0;
+    }
+    
+    ~btDbvntNode()
+    {
+        if (childs[0])
```
[the remark this answers is not in the document]
Comment (xhan0619) [reply src/BulletCollision/BroadphaseCollision/btDbvt.h]: I could change it to more like how btDbvtNode is allocated/deleted. It seems like btDbvtNode is also allocating memory on the fly. See 
static DBVT_INLINE btDbvtNode* createnode(btDbvt* pdbvt,
										  btDbvtNode* parent,
										  void* data)

File: src/BulletSoftBody/btDeformableBodySolver.cpp (hunk 4 above)
Comment (xhan0619) [reply src/BulletSoftBody/btDeformableBodySolver.cpp]: Done.

File: src/BulletSoftBody/btDeformableBodySolver.cpp (hunk 3 above)
Comment (xhan0619) [reply src/BulletSoftBody/btDeformableBodySolver.cpp]: Done.

**Comment by erwincoumans** [comment]
Thanks a lot for the PR. Let's merge it now and we can improve things from here.

**Pull request merged**